src/opam.nix

```diff
@@ -382,6 +382,77 @@ in rec {
       (applyOverlays overlays)
     ];
 
+  # takes an atribute set of package definitions (as produced by `queryToDefs`),
+  # deduplicates sources, and provides a list of sources to fetch
+  defsToSrcs = defs:
+    let
+      urlToName = url:
+        let
+          components = lib.strings.splitString "/" url;
+          filename = lib.last components;
+          name = builtins.head (lib.strings.splitString "." filename);
+        in name;
+      defToSrc = { version, ... }@pkgdef:
+        let
+          inherit (getUrl bootstrapPackages pkgdef) src;
+          name = if pkgdef ? dev-repo then urlToName pkgdef.dev-repo else "";
```

Comment:
Is having `""` as the fallback `name` for the case when there's no `dev-repo` the same as opam-monorepo?

Reply:
opam-monorepo filters out 'virtual' packages (without a dev-repo or url) here: https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/lockfile.ml#L104

I'm still thinking about the best way of doing this with opam-nix.

This isn't the only filtering opam-monorepo does: it also filters packages that aren't build with dune (which is the reason for [opam-overlays](https://github.com/dune-universe/opam-overlays)). But AFAIK we can't do this with `opam admin list`. Maybe modifying the repo to remove any packages not built with dune is the best way to go.

I'm not sure if we need to filter packages build with dune if out of this holds true:
> Package versions from earlier repositories take precedence
over package versions from later repositories.

But if this could take a higher version from a later repository (if the higher version doesn't have an overlay yet) it would cause trouble.

Reply:
I've implemented this [here](https://github.com/tweag/opam-nix/pull/18/commits/5ac390bff6a7c144ef2164739e5f5b58d0041a82), but unfortunately it doesn't consider the case when the newer version *does* build with dune but an older version (which is overlayed) doesn't.

For example, [bheap.1.0.0](https://opam.ocaml.org/packages/bheap/bheap.1.0.0/) doesn't build with dune and therefore has an overlay, but [bheapversion.2.0.0](https://opam.ocaml.org/packages/bheap/bheap.2.0.0/) does build with dune so doesn't have an overlay.

When building [mirage-hello](https://github.com/RyanGibb/mirage-hello) using `opamRepositoryModuloOverlay` we get the error
```
       > [ERROR] No solution including optional dependencies for hello.dev:   * Missing dependency:
       >             - bheap >= 2.0.0
       >             no matching version
```

I think this is why `opam-monorepo` filters (if it depends on dune) by package version rather than package name.

Reply:
One option I can see is parsing every package in the repo and filtering out only those that don't depend on dune. But I imagine this would be quite expensive (even if it would only have to be done once).

Another option would be to use `opam monorepo lock` to resolve dependency versions.